src/coreclr/src/gc/unix/cgroup.cpp

```diff
@@ -59,96 +67,116 @@ class CGroup
+            return GetCGroupMemoryUsage(val, CGROUP1_MEMORY_USAGE_FILENAME);
+        else if (s_cgroup_version == 2)
+            return GetCGroupMemoryUsage(val, CGROUP2_MEMORY_USAGE_FILENAME);
+        else
+        {
+            assert(!"Unknown cgroup version.");
+            return false;
+        }
+    }
 
-        if (s_memory_cgroup_path == nullptr)
-            return result;
+    static bool GetCpuLimit(uint32_t *val)
+    {
+        if (s_cgroup_version == 0)
+            return false;
+        else if (s_cgroup_version == 1)
+            return GetCGroup1CpuLimit(val);
+        else if (s_cgroup_version == 2)
+            return GetCGroup2CpuLimit(val);
+        else
+        {
+            assert(!"Unknown cgroup version.");
+            return false;
+        }
+    }
 
-        size_t len = strlen(s_memory_cgroup_path);
-        len += strlen(MEM_USAGE_FILENAME);
-        mem_usage_filename = (char*)malloc(len+1);
-        if (mem_usage_filename == nullptr)
-            return result;
+private:
+    static int FindCGroupVersion()
+    {
+        int version = 0;
+        char *line = nullptr;
+        size_t lineLen = 0;
+        char *filesystemType = nullptr;
 
-        strcpy(mem_usage_filename, s_memory_cgroup_path);
-        strcat(mem_usage_filename, MEM_USAGE_FILENAME);
-        result = ReadMemoryValueFromFile(mem_usage_filename, &temp);
-        if (result)
+        FILE *mountinfofile = fopen(PROC_MOUNTINFO_FILENAME, "r");
+        if (mountinfofile == nullptr)
+            goto done;
+
+        while (getline(&line, &lineLen, mountinfofile) != -1)
         {
-            if (temp > std::numeric_limits<size_t>::max())
+            // See man page of proc to get format for /proc/self/mountinfo file
+            char *separatorChar = strstr(line, " - ");
+            if (separatorChar == nullptr)
             {
-                *val = std::numeric_limits<size_t>::max();
+                assert(!"Failed to parse mount info file contents.");
+                goto done;
             }
-            else
+            char *filesystemTypeStart = separatorChar + strlen(" - ");
+            if (filesystemTypeStart == nullptr)
             {
-                *val = (size_t)temp;
+                assert(!"Failed to parse mount info file contents.");
+                goto done;
             }
-        }
-        free(mem_usage_filename);
-        return result;
-    }
 
-    static bool GetCpuLimit(uint32_t *val)
-    {
-        long long quota;
-        long long period;
-        double cpu_count;
+            char *filesystemTypeEnd = strstr(filesystemTypeStart, " ");
+            if (filesystemTypeEnd == nullptr)
+            {
+                assert(!"Failed to parse mount info file contents.");
+                goto done;
+            }
 
-        quota = ReadCpuCGroupValue(CFS_QUOTA_FILENAME);
-        if (quota <= 0)
-            return false;
+            size_t filesystemTypeLen = filesystemTypeEnd - filesystemTypeStart;
```

Comment:
Won't this generate a warning because you're assigning a signed expression to an unsigned variable?

Reply:
Hm. For some reason I see no warnings on my end. Will look into this some more.

Reply:
I am going to re-write this entire method based on `statfs` and no actual reading of file contents.